design-documents/remote-invoker.md

```diff
@@ -0,0 +1,631 @@
+## Remote code invoker
+### What for?
+Remotely calling services in proxy versions of modules
+### Overview
+This documents describes remote invoking mechanism, it's functional requirements, the module's API and SPI.
+It also describes a recommended RPC-style networking gateway to call services remotely, but there can be
+multiple _Remote Code Invoker_ implementations based on existing Web APIs, Message Queue RPC API or any other.
+### How are remote services are going to look like with Remote Code Invoker
+\<Module name\>Proxy module will contain \<Module name\>API module's implementations that would look like this:
+```
+class ServiceProxy implements \Magento\ModuleNameApi\ServiceInterface
+{
+
+....
+
+    public function doStuff(\Magento\ModuleNameApi\Data\DTOInterface $dto): \Magento\ModuleNameApi\Data\ResultDTOInterface
+    {
+        $remoteResult = null;
+        $promise = $this->invoker->invoke(
+            'catalog',
+            \Magento\ModuleNameApi\ServiceInterface::class,
+            'doStuff',
+            [$dto]
+        );
+        $promise->then(function (\Magento\ModuleNameApi\Data\ResultDTOInterface $result) use (&$remoteResult) {
+            $remoteResult = $result;
+        });
+        $promise->wait();
```

Comment:
I wonder if we need to have different interfaces for synchronous and asynchronous communication. Developer would know how communication happens. And we will not have to wait for result of asynchronous operation, one of the benefits of async operations is that we don't need to wait for the response and it can be performed quickly.